src/core/thread/key_manager.cpp

```diff
@@ -362,20 +368,13 @@ void KeyManager::UpdateKeyMaterial(void)
 #endif
 }
 
-void KeyManager::SetCurrentKeySequence(uint32_t aKeySequence)
+void KeyManager::SetCurrentKeySequence(uint32_t aKeySequence, KeySequenceUpdateMode aUpdateMode)
 {
     VerifyOrExit(aKeySequence != mKeySequence, Get<Notifier>().SignalIfFirst(kEventThreadKeySeqCounterChanged));
 
-    if ((aKeySequence == (mKeySequence + 1)) && mKeyRotationTimer.IsRunning())
+    if (aUpdateMode == kApplyKeySwitchGuard)
```

Comment:
In line 373, I was wondering why `kEventThreadKeySeqCounterChanged` is signalled in case the tKSC is not actually changed, but already equal to the desired value?

Reply:
We use `SignalIfFirst()` in this case:
```c++
    /**
     * Schedules signaling of am event only if the event has not been signaled before (first time signal).
     *
     * @param[in]  aEvent     The event to signal.
     *
     */
    void SignalIfFirst(Event aEvent);
```

This pattern is used for all `Notifier` events to ensure changes are signaled when the OT stack is first initialized (including on reboot, reset, etc.). This guarantees events are signaled even if the new value matches the initial/default value (or whatever value may be in the memory for member variable).